### ISSUE TITLE
New feature - Specify constraints for DataFrames 

### ISSUE THREAD

**Issue by el-oso** [JuliaData/DataFrames.jl#3033]
`DataFrames` is great as an exploration tool, but then when we found something that works I just want to make sure that things are well structured and reproducible. Then is when I like `DataFrames` a bit less, because then I cannot guarantee that a `DataFrame` will have column "a" or "b" due to it's dynamic nature. I want to make sure that an object has the same structure always and I don't want to make those checks at runtime, I prefer to do those check upfront.

Inspired by this article https://towardsdatascience.com/schema-specification-for-your-pandas-dataframes-c8b34d1eb37d I 

@bkamins suggested: "I guess what you want is that you could specify some constraints your data frame must meet (as I read the pandas docs the main benefit seems that you can say that I want columns a, b and c have types x, y and z, but at the same time you allow the data frame to have additional columns - is this right)? We potentially could (but we need to discuss this if this makes sense) allow the user to register a validator function with a data frame that would be run on every operation that potentially might change a schema of DataFrame. I guess something like this would solve your problem - right?"

This sounds like a potentially good solution and an addition for `DataFrames`.

We also discussed about `TypedTables.jl` suggested by @pdeffebach , but it is not a drop in replacement.

**Comment by el-oso** [comment]
Yes, indeed. I just added it as an extra remark to summarize our discussion on Slack and put it in context.
I like the validator option as you suggested.

**Comment by el-oso** [comment]
A predicate would be a good start indeed, but how do I dispatch `DataFrame`  with predicate A and `DataFrame` with predicate B? With a parameter? or perhaps use holy traits?

**Comment by bkamins** [comment]
What you can do currently:
1. register a predicate as a `metadata` field, e.g. call it `"validator"` that would have `:note` style.
2. write a simple function e.g.:
```
v(df) = (metadata(df, "validator")(df); df)
```
and wherever you use `df`, use `v(df)` instead. It is not more typing than in typed pandas where you need to unwrap `df` as a field.

Considerations:
1. due to design of Julia you can always mutate `DataFrame` object manually, so there is no way to ensure the predicate always holds without manual call to a function checking it.
2. adding this check to every operation that mutates `DataFrame` in functions defined in DataFrames.jl would kill performance.
3. We could add a check in some specific places in the code (before expensive operations) as then the cost would be negligible, but I am not sure how useful users would find it (currently we run such check to make sure that data frame is not corrupted).

**Comment by el-oso** [comment]
What I don't like about this approach is that is still a check on runtime. I would rather have a compile time check or at the very least ensure that `DataFrameA`   (that is a `DataFrame` with schema `A`) is called and not `DataFrameB`

**Comment by bkamins** [comment]
You could subtype `AbstractDataFrame` with fixed schema.
